src/Form/GroupUserSubscribeForm.php

```diff
@@ -0,0 +1,129 @@
+    if ($group->access('view')) {
+      return new Url("entity.{$group->getEntityTypeId()}.og_admin_routes.members", [$group->getEntityTypeId() => $group->id()]);
+    }
+    return new Url('<front>');
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @see \Drupal\Core\Entity\EntityConfirmFormBase::buildForm
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildForm($form, $form_state);
+    /** @var \Drupal\og\OgMembershipInterface $membership */
+    $membership = $this->entity;
+    $form['uid'] = [
+      '#type' => 'entity_autocomplete',
+      '#title' => $this->t('Member'),
+      '#target_type' => 'user',
+    ];
+    $options = [];
+    /** @var \Drupal\og\OgRoleInterface $role */
+    foreach (OgRole::loadByGroupType($membership->getGroup()->getEntityTypeId(), $membership->getGroup()->bundle()) as $role) {
+      // Only add the role to the list if it is not a required role, these
+      // cannot be added. Nor should invalid roles be added.
+      if (!$role->isRequired() && $membership->isRoleValid($role)) {
+        $options[$role->id()] = $role->label();
+      }
+    }
+    $form['roles'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Roles'),
+      '#multiple' => TRUE,
+      '#options' => $options,
+    ];
+    $form[OgMembershipInterface::REQUEST_FIELD]['#access'] = FALSE;
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function actions(array $form, FormStateInterface $form_state) {
+    $actions = parent::actions($form, $form_state);
+    $actions['submit']['#value'] = $this->t('Add');
+    $actions['cancel'] = [
+      '#type' => 'link',
+      '#title' => $this->t('Cancel'),
+      '#attributes' => ['class' => ['button']],
+      '#url' => $this->getCancelUrl(),
+      '#cache' => [
+        'contexts' => [
+          'url.query_args:destination',
+        ],
+      ],
+    ];
+    return $actions;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+    /** @var \Drupal\og\OgMembershipInterface $membership */
+    $membership = $this->entity;
+    $result = $this->entityTypeManager
+      ->getStorage($membership->getEntityTypeId())
+      ->getQuery()
+      ->condition('uid', $form_state->getValue('uid'))
+      ->condition('type', $membership->bundle())
+      ->condition('entity_type', $membership->getGroup()->getEntityTypeId())
+      ->condition('entity_id', $membership->getGroup()->id())
+      ->execute();
+    if ($result) {
+      $form_state->setErrorByName('uid', $this->t('The user is already a member of the group.'));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+    /** @var \Drupal\og\OgMembershipInterface $membership */
+    $membership = $this->entity;
```

Comment:
Thanks for the useful PR! I tried it and the roles aren't getting saved. So we need to save the roles here:

```suggestion
    $membership = $this->entity;
    $assigned_roles = $form_state->getValue('roles');
    if (!empty($assigned_roles)) {
      $roles = OgRole::loadMultiple($assigned_roles);
      foreach($roles as $role) {
        $membership->addRole($role);
      }
    }
```

Reply:
Not sure what went wrong here but roles is a basefield on the membership entity so it should get saved.
```
    $fields['roles'] = BaseFieldDefinition::create('entity_reference')
      ->setLabel(t('Roles'))
      ->setDescription(t('The OG roles related to an OG membership entity.'))
      ->setCardinality(FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED)
      ->setSetting('target_type', 'og_role')
      ->setSetting('handler', 'og:og_role')
      ->setConstraints(['ValidOgRole' => []])
      ->setDisplayOptions('form', [
        'type' => 'options_buttons',
        'weight' => 0,
      ])
      ->setDisplayConfigurable('view', TRUE)
      ->setDisplayConfigurable('form', TRUE);
```

Are you on the dev branch?

Reply:
I was wondering about that, I'm on `1.0.0-alpha5`. If it's already resolved false alarm then! Thanks